File: QuizGame.cpp
#include "QuizGame.h"
#include <iostream>
#include <random>
#include <algorithm>
#include <iomanip>

QuizGame::QuizGame(std::map<std::string, int> e)
    : MiniGame("Викторина", e), difficulty(5), correctAnswers(0) {
    loadQuestions();
}

void QuizGame::loadQuestions() {
    allQuestions.push_back({
        "Кто сказал: Ребята, давайте жить дружно!",
        {"Волк", "Лиса", "Кот Леопольд", "Матроскин"},
        2, "мультики"
        });
    allQuestions.push_back({
        "Как звали кота из Простоквашино?",
        {"Барсик", "Матроскин", "Мурзик", "Феликс"},
        1, "мультики"
        });
    allQuestions.push_back({
        "Кто такой Чебурашка?",
        {"Обезьяна", "Неизвестный зверь", "Медведь", "Енот"},
        1, "мультики"
        });
    allQuestions.push_back({
        "Какого цвета был Волк из Ну погоди?",
        {"Серый", "Белый", "Чёрный", "Рыжий"},
        0, "мультики"
        });
    allQuestions.push_back({
        "Кто жил в Изумрудном городе с Элли?",
        {"Тотошка", "Бим", "Бобик", "Шарик"},
        0, "мультики"
        });
    allQuestions.push_back({
        "Что такое std::vector?",
        {"Массив", "Динамический массив", "Список", "Очередь"},
        1, "cpp"
        });
    allQuestions.push_back({
        "Какой оператор для доступа через указатель?",
        {".", "->", "::", "*"},
        1, "cpp"
        });
    allQuestions.push_back({
        "Что возвращает main() при успехе?",
        {"0", "1", "-1", "nullptr"},
        0, "cpp"
        });
    allQuestions.push_back({
        "Как объявить константу в C++?",
        {"const", "final", "static", "define"},
        0, "cpp"
        });
    allQuestions.push_back({
        "Что такое unique_ptr?",
        {"Умный указатель", "Обычный указатель", "Массив", "Функция"},
        0, "cpp"
        });
}

void QuizGame::selectQuestions() {
    std::random_device rd;
    std::mt19937 g(rd());
    std::shuffle(allQuestions.begin(), allQuestions.end(), g);

    questions.clear();
    int count = std::min(difficulty, (int)allQuestions.size());
    for (int i = 0; i < count; i++) {
        questions.push_back(allQuestions[i]);
    }
}

bool QuizGame::play(PetBase& pet) {
    std::cout << "\n";
    std::cout << "********************************************************\n";
    std::cout << "*                    ВИКТОРИНА                         *\n";
    std::cout << "********************************************************\n";
    std::cout << "Выберите сложность:\n";
    std::cout << "  1. Легкая (3 вопроса)\n";
    std::cout << "  2. Средняя (5 вопросов)\n";
    std::cout << "  3. Сложная (7 вопросов)\n";
    std::cout << "Ваш выбор: ";

    int diffChoice;
    std::cin >> diffChoice;
    difficulty = (diffChoice == 1) ? 3 : (diffChoice == 3) ? 7 : 5;

    selectQuestions();
    correctAnswers = 0;

    std::cout << "\nНачинаем! Отвечайте цифрой (1-4)\n";

    for (size_t i = 0; i < questions.size(); ++i) {
        const auto& q = questions[i];
        std::cout << "\n--------------------------------------------------------\n";
        std::cout << "[" << (q.category == "cpp" ? "C++" : "Мультфильмы") << "] ";
        std::cout << "Вопрос " << (i + 1) << "/" << questions.size() << "\n";
        std::cout << "--------------------------------------------------------\n";
        std::cout << q.text << "\n\n";

        for (size_t j = 0; j < q.options.size(); ++j) {
            std::cout << "  " << (j + 1) << ". " << q.options[j] << "\n";
        }

        int answer;
        std::cout << "\nВаш ответ: ";
        std::cin >> answer;
        if (answer == q.correct + 1) {
            correctAnswers++;
            std::cout << "Правильно!\n";
        }
        else {
            std::cout << "Неправильно! Правильный ответ: " << (q.correct + 1) << "\n";
        }
    }

    bool win = (correctAnswers >= questions.size() / 2);

    // ПОКАЗАТЬ РЕЗУЛЬТАТЫ
    std::cout << "\n";
    std::cout << "********************************************************\n";
    std::cout << "*                   РЕЗУЛЬТАТЫ                         *\n";
    std::cout << "********************************************************\n";
    std::cout << "*  Правильных ответов: " << std::setw(3) << correctAnswers << " из " << questions.size() << "              *\n";
    std::cout << "*  Процент успеха:     " << std::setw(3) << (correctAnswers * 100 / questions.size()) << "%                      *\n";
    std::cout << "*  " << (win ? "ПОБЕДА!" : "Попробуйте еще раз") << "                              *\n";
    std::cout << "********************************************************\n";

    int moodBonus = (correctAnswers * 10);
    pet.getParameters().setMood(pet.getParameters().getMood() + moodBonus);
    pet.getParameters().setFatigue(pet.getParameters().getFatigue() + 5);

    // ПАУЗА
    std::cout << "\nНажмите Enter для продолжения...";
    std::cin.ignore(10000, '\n');
    std::cin.get();

    return win;
}

void QuizGame::showResults() const {
    std::cout << "\nСтатистика викторины: " << correctAnswers << " правильных ответов\n";
}

File: QuizGame.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Question
{
    public string Text { get; set; }
    public List<string> Options { get; set; }
    public int Correct { get; set; }
    public string Category { get; set; }
}

public class QuizGame : MiniGame
{
    private List<Question> questions;
    private List<Question> allQuestions;
    private int difficulty;
    private int correctAnswers;

    public QuizGame(Dictionary<string, int> effects) : base("Викторина", effects)
    {
        allQuestions = new List<Question>();
        questions = new List<Question>();
        difficulty = 5;
        correctAnswers = 0;
        LoadQuestions();
    }

    private void LoadQuestions()
    {
        allQuestions.Add(new Question
        {
            Text = "Кто сказал: Ребята, давайте жить дружно!",
            Options = new List<string> { "Волк", "Лиса", "Кот Леопольд", "Матроскин" },
            Correct = 2,
            Category = "мультики"
        });

        allQuestions.Add(new Question
        {
            Text = "Как звали кота из Простоквашино?",
            Options = new List<string> { "Барсик", "Матроскин", "Мурзик", "Феликс" },
            Correct = 1,
            Category = "мультики"
        });

        allQuestions.Add(new Question
        {
            Text = "Что такое std::vector?",
            Options = new List<string> { "Массив", "Динамический массив", "Список", "Очередь" },
            Correct = 1,
            Category = "cpp"
        });

        allQuestions.Add(new Question
        {
            Text = "Какой оператор для доступа через указатель?",
            Options = new List<string> { ".", "->", "::", "*" },
            Correct = 1,
            Category = "cpp"
        });

        allQuestions.Add(new Question
        {
            Text = "Что возвращает main() при успехе?",
            Options = new List<string> { "0", "1", "-1", "nullptr" },
            Correct = 0,
            Category = "cpp"
        });
    }

    private void SelectQuestions()
    {
        var random = new Random();
        var shuffled = allQuestions.OrderBy(x => random.Next()).Take(difficulty).ToList();
        questions = shuffled;
    }

    public override bool Play(PetBase pet)
    {
        Console.WriteLine("\n********************************************************");
        Console.WriteLine("*                    ВИКТОРИНА                         *");
        Console.WriteLine("********************************************************");
        Console.WriteLine("Выберите сложность:");
        Console.WriteLine("  1. Легкая (3 вопроса)");
        Console.WriteLine("  2. Средняя (5 вопросов)");
        Console.WriteLine("  3. Сложная (7 вопросов)");
        Console.Write("Ваш выбор: ");

        int diffChoice = int.Parse(Console.ReadLine() ?? "2");
        difficulty = diffChoice == 1 ? 3 : diffChoice == 3 ? 7 : 5;

        SelectQuestions();
        correctAnswers = 0;

        Console.WriteLine("\nНачинаем! Отвечайте цифрой (1-4)\n");

        for (int i = 0; i < questions.Count; i++)
        {
            var q = questions[i];
            Console.WriteLine($"--------------------------------------------------------");
            Console.WriteLine($"[{(q.Category == "cpp" ? "C++" : "Мультфильмы")}] Вопрос {i + 1}/{questions.Count}");
            Console.WriteLine($"--------------------------------------------------------");
            Console.WriteLine(q.Text);
            Console.WriteLine();

            for (int j = 0; j < q.Options.Count; j++)
            {
                Console.WriteLine($"  {j + 1}. {q.Options[j]}");
            }

            Console.Write("\nВаш ответ: ");
            int answer = int.Parse(Console.ReadLine() ?? "0");

            if (answer == q.Correct + 1)
            {
                correctAnswers++;
                Console.WriteLine("Правильно!");
            }
            else
            {
                Console.WriteLine($"Неправильно! Правильный ответ: {q.Correct + 1}");
            }
        }

        bool win = correctAnswers >= questions.Count / 2;

        Console.WriteLine("\n╔════════════════════════════════════════════════════════╗");
        Console.WriteLine("║                   РЕЗУЛЬТАТЫ                           ║");
        Console.WriteLine("╠════════════════════════════════════════════════════════╣");
        Console.WriteLine($"║  Правильных ответов: {correctAnswers,3} из {questions.Count,-12}               ");
        int percentage = questions.Count > 0 ? correctAnswers * 100 / questions.Count : 0;
        Console.WriteLine($"║  Процент успеха:     {percentage,3}%                              ");
        string resultText = win ? "ПОБЕДА!" : "Попробуйте еще раз";
        Console.WriteLine($"║  {resultText,-52}  ");
        Console.WriteLine("╚════════════════════════════════════════════════════════╝");

        int moodBonus = correctAnswers * 10;
        pet.GetParameters().SetMood(pet.GetParameters().GetMood() + moodBonus);
        pet.GetParameters().SetFatigue(pet.GetParameters().GetFatigue() + 5);

        Console.WriteLine("\nНажмите Enter для продолжения...");
        Console.ReadLine();

        return win;
    }
}